File: ml/rtsml/protocol.py
"""The frame protocol on the pipe to a Bun environment process.

Mirrors `tools/ml/protocol.ts`: `u32 length | u8 kind | u32 headerLength |
header JSON | payload`, little-endian, with the header's `arrays` naming the
payload's arrays in order.
"""

from __future__ import annotations

import json
import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import Any, BinaryIO

import numpy as np


class Kind(IntEnum):
    HELLO = 1
    RESET = 2
    STEP = 3
    OBS = 4
    ERROR = 5
    CLOSE = 6


DTYPES = {"f32": np.dtype("<f4"), "u8": np.dtype("u1"), "i32": np.dtype("<i4")}
NAMES = {np.dtype("float32"): "f32", np.dtype("uint8"): "u8", np.dtype("int32"): "i32"}


@dataclass
class Frame:
    kind: Kind
    header: dict[str, Any]
    arrays: list[np.ndarray]


def encode_frame(kind: Kind, header: dict[str, Any], arrays: list[tuple[str, np.ndarray]] | None = None) -> bytes:
    arrays = arrays or []
    descs = []
    payload = bytearray()
    for name, array in arrays:
        array = np.ascontiguousarray(array)
        if array.dtype not in NAMES:
            raise TypeError(f"unsupported dtype {array.dtype} for {name}")
        descs.append({"name": name, "dtype": NAMES[array.dtype], "shape": list(array.shape)})
        payload += array.astype(DTYPES[NAMES[array.dtype]], copy=False).tobytes()
    header_bytes = json.dumps({**header, "arrays": descs}).encode("utf-8")
    length = 1 + 4 + len(header_bytes) + len(payload)
    return struct.pack("<IBI", length, int(kind), len(header_bytes)) + header_bytes + bytes(payload)


def decode_frame(data: bytes) -> Frame:
    length, kind, header_len = struct.unpack_from("<IBI", data, 0)
    if len(data) != 4 + length:
        raise ValueError(f"frame is {len(data)} bytes, header says {4 + length}")
    header = json.loads(data[9 : 9 + header_len].decode("utf-8"))
    arrays = []
    at = 9 + header_len
    for desc in header.get("arrays", []):
        dtype = DTYPES[desc["dtype"]]
        count = int(np.prod(desc["shape"])) if desc["shape"] else 1
        size = count * dtype.itemsize
        arrays.append(np.frombuffer(data[at : at + size], dtype=dtype).reshape(desc["shape"]))
        at += size
    if at != len(data):
        raise ValueError(f"frame has {len(data) - at} trailing bytes")
    return Frame(Kind(kind), header, arrays)


def read_frame(stream: BinaryIO) -> Frame:
    head = _read_exactly(stream, 4)
    (length,) = struct.unpack("<I", head)
    return decode_frame(head + _read_exactly(stream, length))


def write_frame(stream: BinaryIO, frame: bytes) -> None:
    stream.write(frame)
    stream.flush()


def _read_exactly(stream: BinaryIO, n: int) -> bytes:
    out = bytearray()
    while len(out) < n:
        chunk = stream.read(n - len(out))
        if not chunk:
            raise EOFError("environment process closed the pipe")
        out += chunk
    return bytes(out)

File: tools/ml/protocol.ts
/**
 * The wire between Python and a Bun environment process.
 *
 * Frames on stdin and stdout, each:
 *
 *   u32 length (of everything after it) | u8 kind | u32 headerLength |
 *   header JSON | payload
 *
 * little-endian. The header names the arrays in the payload, in order, with
 * their dtype and shape, so the payload is raw bytes with no framing of its
 * own — a step's observations for a dozen environments is one frame and one
 * write. `rtsml/protocol.py` is the same layout in Python.
 */

export enum Kind {
  Hello = 1,
  Reset = 2,
  Step = 3,
  Obs = 4,
  Error = 5,
  Close = 6,
}

export type DType = 'f32' | 'u8' | 'i32';

export interface ArrayDesc {
  name: string;
  dtype: DType;
  shape: number[];
}

export interface Frame<H = unknown> {
  kind: Kind;
  header: H;
  arrays: ArrayBufferView[];
}

const encoder = new TextEncoder();
const decoder = new TextDecoder();

function dtypeOf(view: ArrayBufferView): DType {
  if (view instanceof Float32Array) return 'f32';
  if (view instanceof Uint8Array) return 'u8';
  if (view instanceof Int32Array) return 'i32';
  throw new Error('unsupported array type');
}

function bytesOf(view: ArrayBufferView): Uint8Array {
  return new Uint8Array(view.buffer, view.byteOffset, view.byteLength);
}

/** Serialise a frame. `arrays` are listed in the header as `arrays: ArrayDesc[]`. */
export function encodeFrame(
  kind: Kind,
  header: Record<string, unknown>,
  arrays: readonly { name: string; view: ArrayBufferView; shape?: number[] }[] = [],
): Uint8Array {
  const descs: ArrayDesc[] = arrays.map((a) => ({
    name: a.name,
    dtype: dtypeOf(a.view),
    shape: a.shape ?? [a.view.byteLength / bytesPer(dtypeOf(a.view))],
  }));
  const headerBytes = encoder.encode(JSON.stringify({ ...header, arrays: descs }));
  let payload = 0;
  for (const a of arrays) payload += a.view.byteLength;
  const length = 1 + 4 + headerBytes.length + payload;
  const out = new Uint8Array(4 + length);
  const dv = new DataView(out.buffer);
  dv.setUint32(0, length, true);
  out[4] = kind;
  dv.setUint32(5, headerBytes.length, true);
  out.set(headerBytes, 9);
  let at = 9 + headerBytes.length;
  for (const a of arrays) {
    out.set(bytesOf(a.view), at);
    at += a.view.byteLength;
  }
  return out;
}

export function bytesPer(dtype: DType): number {
  return dtype === 'u8' ? 1 : 4;
}

function viewOf(dtype: DType, bytes: Uint8Array): ArrayBufferView {
  // Copy into a fresh, aligned buffer: a slice of the stream is not aligned.
  const copy = new Uint8Array(bytes.length);
  copy.set(bytes);
  switch (dtype) {
    case 'f32':
      return new Float32Array(copy.buffer);
    case 'i32':
      return new Int32Array(copy.buffer);
    case 'u8':
      return copy;
  }
}

/** Parse one complete frame from `bytes` (which must be exactly one frame). */
export function decodeFrame<H = Record<string, unknown>>(
  bytes: Uint8Array,
): Frame<H & { arrays: ArrayDesc[] }> {
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const length = dv.getUint32(0, true);
  if (bytes.length !== 4 + length)
    throw new Error(`frame is ${bytes.length} bytes, header says ${4 + length}`);
  const kind = bytes[4] as Kind;
  const headerLength = dv.getUint32(5, true);
  const header = JSON.parse(decoder.decode(bytes.subarray(9, 9 + headerLength))) as H & {
    arrays: ArrayDesc[];
  };
  const arrays: ArrayBufferView[] = [];
  let at = 9 + headerLength;
  for (const desc of header.arrays ?? []) {
    const count = desc.shape.reduce((n, d) => n * d, 1);
    const size = count * bytesPer(desc.dtype);
    arrays.push(viewOf(desc.dtype, bytes.subarray(at, at + size)));
    at += size;
  }
  if (at !== bytes.length) throw new Error(`frame has ${bytes.length - at} trailing bytes`);
  return { kind, header, arrays };
}

/** Feeds arbitrary chunks and yields complete frames, however the bytes were split. */
export class FrameParser {
  private buffer = new Uint8Array(0);

  push(chunk: Uint8Array): Uint8Array[] {
    const joined = new Uint8Array(this.buffer.length + chunk.length);
    joined.set(this.buffer);
    joined.set(chunk, this.buffer.length);
    const frames: Uint8Array[] = [];
    let at = 0;
    while (joined.length - at >= 4) {
      const length = new DataView(joined.buffer, joined.byteOffset + at, 4).getUint32(0, true);
      if (joined.length - at < 4 + length) break;
      frames.push(joined.slice(at, at + 4 + length));
      at += 4 + length;
    }
    this.buffer = joined.slice(at);
    return frames;
  }
}
